OAuth

voidxnull opened the pull request:
Hey there. This is my first PR for the project so I'll probably need your help to polish it.

The OAuth flow goes like this:
1. Client opens `/v2/oauth/[provider]` in a secondary window. This route redirects to the provider page.
2. After a user is authenticated/authorized, the provider page redirects to `/v2/oauth/[provider]/callback` which responds with a script (https://github.com/CryptodelicUK/freefeed-server/blob/79c537a4e75c15866e22ec5bff2a07200974f13f/config/initializers/passport.js#L130). 
3. The script posts a message to the opener and closes the secondary window. The message contains either `authToken` or `error`.
4. Opener processes the message (https://github.com/CryptodelicUK/freefeed-react-client/blob/c6b6629444f91157f2377c8a6bc742f3b9b73fa8/src/services/auth.js#L27).

Since we can't rely on the OAuth2 `state` parameter (some providers may use OAuth1) I used cookies for storing `freefeed_authToken` and `referer` for callbacks.

Also added authorization flow for facebook:
`/v2/oauth/facebook/authz` - Redirects to the provider page.
`/v2/oauth/facebook/authz/callback` - Additional callback url should be added in facebook settings.
`/v2/oauth/facebook/friends` - Fetches friends through the graph API, transforms them into freefeed users. Uses the cached access token `users.providers.facebook.accessToken` if not specified explicitly as a query param. 

### Configuration
#### Provider keys:
```
config.oauth = {
    facebookClientId:     'key',
    facebookClientSecret: 'secret',
    googleClientId:       'key',
    googleClientSecret:   'secret',
    githubClientId:       'key',
    githubClientSecret:   'secret',
  };
```
#### `authToken` cookie prefix (to store user session for OAuth callbacks):
```
config.authTokenPrefix: 'freefeed_',
```

voidxnull committed:
Add tests

- unit tests for generateUsername()
- unit tests for renderCallbackResponse()
- unit tests for auth method User methods
- unit tests for getAuthParams and getAuthzParams
- integration tests for createUserFromOauth()
- functional tests for OauthController
- functional tests for authorize() and authorizeCallback()
- functional tests for facebook friends controllers

davidmz commented on app/routes/api/v2/OauthRoute.js:
```diff
@@ -0,0 +1,24 @@
+export default function addRoutes(app) {
+  if (!isEmpty(oauthConfig.facebookClientId)) {
+    app.get('/v2/oauth/facebook', OauthController.facebook);
+    app.get('/v2/oauth/facebook/callback', OauthController.facebookCallback);
```
Probably it is better to use common patterns like '/v2/oauth/:provider' and '/v2/oauth/:provider/callback' here and check config in controller methods?

voidxnull replied:
This will require a refactoring. Working on it.

voidxnull replied:
Done.

davidmz commented on config/environments/console.js:
```diff
@@ -113,5 +113,14 @@ export function getConfig() {
+    githubClientSecret:   '',
+  };
+
   return config;
```
There change should also be applied to the 'config/environments/test.js' file

voidxnull replied:
*Done.*

davidmz commented:
I think, user should have the ability to change their OAuth bindings and to unbind some or all of these. In the last case user should still be able to authorize by password.

davidmz commented on app/models/user.js:
```diff
@@ -324,15 +325,18 @@ export function addModel(dbAdapter) {
+
+    if (!skip_password) {
+      await this.initPassword()
+    }
```
It would be better to generate a long random password than leave it empty (null). We have internal rule that the empty string in 'hashed_password' field means blocked account. Adding another special meaning value (null) will lead to overcomplication.

voidxnull replied:
Ok, I'll change that.

davidmz commented on config/initializers/passport.js:
```diff
@@ -29,5 +38,225 @@ export function init(passport) {
+    <script>
+      var error = ${JSON.stringify(error && error.message)};
+      var authToken = ${JSON.stringify(authToken)};
+      window.opener.postMessage({ error: error, authToken: authToken }, '*');
```
Is it safe to use '*' here?

voidxnull replied:
Yeah, this is my oversight. Should have replaced it with `config.origin`. *Done.*

davidmz replied:
In the real world (https://github.com/FreeFeed/freefeed-ansible/blob/6bbcad420c5bdcf167c6e8ad08c8cea3b2c6d52f/roles/pepyatka/defaults/main.yml#L11) the config.origin is actually a '*':) So may be we need to get origin from client before the start of authorization process?

voidxnull replied:
Done. I used cookies to temporarily store the origin.

voidxnull commented:
I'm sorry for the delay, I was quite sick last week.

It seems like I addressed all the comments except for provider binding/unbinding. Here's a short plan for implementing it:

### Server:
- [x] Controller for attaching a new provider to existing accounts.
- [x] Controller for detaching a provider.
- [x] Make sure `MyProfileSerializer` includes `users.providers`.

### Client:
- [x] Add API entries to `services/api.js`.
- [x] Add Redux actions/reducers.
- [x] Component for editing attached providers.
  - [x] Displays providers as big buttons with icons.
  - [x] It must highlight attached providers and dim unattached.
  - [x] Click on an attached provider icon opens a dialog asking to detach the provider. (show a cross on hover?)
  - [x] Click on an unattached provider icon opens the provider authentication page.
  - [x] Add the component to the user settings page.
- [ ] ...

voidxnull commented:
Done. Also added facebook authz and friend list API to the PR.

voidxnull commented:
Refactored to support the ability to add multiple external accounts of the same provider.

voidxnull committed:
Add OAuth support

- Added migration which adds `providers` column to `users`.
- Added new passport strategies for Facebook, Google, and Github.
- Added OAuth controllers (OauthController).
- Retrieve authToken from cookies.
- Implement OAuth authorization and friend list for facebook.
- Implement ability to add multiple auth methods.

indeyets commented:
implemented via #403
